include/envoy/http/filter.h

```diff
@@ -441,6 +449,13 @@ class StreamEncoderFilterCallbacks : public virtual StreamFilterCallbacks {
    */
   virtual HeaderMap& addEncodedTrailers() PURE;
 
+  /**
+   * Adds encoded metadata.
```

Comment:
Can we discuss how this works, and extend the docs to add some clarity?

As I read it it's not always adding metadata - sometimes it's adding one but often it's returning a reference to an already existing metadata frame, depending on if the HCM has had a chance to drain.
I'm wondering if the lifetime of metadata frames would be easier to understand if we either represented this as a list (where add would actually add) or did what we do on incoming metadata and immediately shove it through the filter chain and out to the peer?

Reply:
Sorry I didn't explain the change either in the description or in the doc!

You are right! I was mistakenly assuming a user doesn't add the same metadata multiple times. That is, if a metadata is added in encodeHeader(), then it doesn't need to be added again in encodeData(). As a result, different metadata can be stored in the same metadata map, since they are different.

That's a wrong and unnecessary assumption. I also realized we need to have the ability to encode a list of metadata map in the request direction anyway (because we may accumulate multiple metadata maps while waiting for the connection pool to be ready, unlike the response direction, where we can shove it through immediately.)

I am working on a PR to enable encoding a list of metadata_map. After the PR is submitted, I will update this PR, and store response_metadata_map_ in a list as suggested. 

Sorry about the iteration!